### PR TITLE
Syncing new test vectors and base mode

### PR DESCRIPTION
**Pull request by kevinlewi**
To sync with version 09 of voprf. This version introduces a new mode (so we have three now: base, verifiable, and "poprf").

### CODE REVIEW

**Comment by kevinlewi** [comment]
@daxpedda : In the latest version, only the "POPRF" mode uses the "PreparedTScalar" and will follow the batch_evaluate_prepare -> batch_evaluate_finish paradigm.

However, the new "Verifiable" mode no longer needs to create a PreparedTScalar, and can just have a single function execute the batch_evaluate, I believe. I was hoping to make this change, but was struggling to make sense of some of the types that were introduced here. Happy to sync off-line if you get a chance and I can explain further where I got stuck!

**Comment by daxpedda** [comment]
Will take a look soon'ish!

File: src/voprf.rs [hunk 1]
```diff
@@ -908,6 +833,46 @@ where
...
+    let context_string = create_context_string::<CS>(mode);
+    let dst = GenericArray::from(STR_DERIVE_KEYPAIR).concat(context_string);
+
+    // deriveInput = seed || I2OSP(len(info), 2) || info
+    let info_len = i2osp_2(info.len()).map_err(|_| Error::Metadata)?;
+
+    let mut counter: usize = 0;
+
+    loop {
+        if counter > 255 {
```
Comment (daxpedda) [inline src/voprf.rs:855]: This is probably better in a `for counter in 0..=255` loop.

File: src/voprf.rs [hunk 2]
```diff
@@ -908,6 +833,46 @@ where
...
+            break Err(Error::DeriveKeyPair);
+        }
+
+        // skS = G.HashToScalar(deriveInput || I2OSP(counter, 1), DST = "DeriveKeyPair"
+        // || contextString)
+        let counter_i2osp = i2osp_1(counter).map_err(|_| Error::DeriveKeyPair)?;
```
Comment (daxpedda) [inline src/voprf.rs:861]: If we let `counter` be a `u8`, we can simply use `.to_be_bytes()` here, which also can't fail.

File: src/voprf.rs [hunk 3]
```diff
@@ -1135,8 +1091,7 @@ where
         &a2,
         &elem_len,
         &a3,
-        &challenge_dst_len,
-        &challenge_dst,
+        &GenericArray::from(STR_CHALLENGE),
```
Comment (daxpedda) [inline src/voprf.rs:1094]: We don't need to use `GenericArray` here, this is already a byte slice.

File: src/voprf.rs [hunk 4]
```diff
@@ -1176,30 +1130,24 @@ where
...
-    let finalize_dst = GenericArray::from(STR_FINALIZE).concat(get_context_string::<CS>(mode));
+    let finalize_dst = GenericArray::from(STR_FINALIZE);
 
     inputs_and_unblinded_elements
         // To make a return type possible, we have to convert to a `fn` pointer,
         // which isn't possible if we `move` from context.
-        .zip(iter::repeat((info, finalize_dst)))
-        .map(|((input, unblinded_element), (info, finalize_dst))| {
-            let finalize_dst_len = i2osp_2_array(&finalize_dst);
+        .zip(iter::repeat(finalize_dst))
```
Comment (daxpedda) [inline src/voprf.rs:1138]: Now that `finalize_dst` is "free" we won't need to pass it through here either.

File: src/voprf.rs [hunk 5]
```diff
@@ -1176,30 +1130,24 @@ where
     // https://www.ietf.org/archive/id/draft-irtf-cfrg-voprf-08.html#section-3.3.3.2-2
     // https://www.ietf.org/archive/id/draft-irtf-cfrg-voprf-08.html#section-3.3.4.3-1
 
-    // finalizeDST = "Finalize-" || contextString
-    let finalize_dst = GenericArray::from(STR_FINALIZE).concat(get_context_string::<CS>(mode));
+    let finalize_dst = GenericArray::from(STR_FINALIZE);
```
Comment (daxpedda) [inline src/voprf.rs:1133]: As above, `STR_FINALIZE` is already a byte slice.

File: src/voprf.rs [hunk 6]
```diff
@@ -1264,9 +1214,9 @@ where
             &ci,
             &elem_len,
             &di,
-            &composite_dst_len,
-            &composite_dst,
+            &GenericArray::from(STR_COMPOSITE),
```
Comment (daxpedda) [inline src/voprf.rs:1217]: As above.

**Review by daxpedda — CHANGES_REQUESTED** [review]
LGTM.

Things I will do tmrw:
- Actually go check with the spec, I only reviewed how the code looks.
- Fix `batch_evaluate`. Will make a PR against this branch.

**Comment by kevinlewi** [comment]
Thanks for leaving the comments! I have not addressed them just yet (am adding POPRF in first), and could use your eyes on that first...!

**Comment by kevinlewi** [comment]
As of the latest commit on this branch, the implementation should be in sync with the CFRG's test vectors for all three modes (OPRF, VOPRF, POPRF). I split up the previous "voprf.rs" into three files, one for each mode. However, I removed the no-alloc versions since I could not get them to compile. I believe the no-alloc version is only needed for the batch functions of POPRF, but not VOPRF.

Would be great if you could help here @daxpedda !

Also, the doctests do not currently pass, and many docs for functions need to be updated. Will be happy to do so once we have the no-alloc version working!

**Comment by daxpedda** [comment]
I addressed the remaining concerns in #59.

**Comment by kevinlewi** [comment]
Merging along with #59

**Pull request merged**